software/contrib/sigma.py

```diff
@@ -0,0 +1,399 @@
+
+    AIN_ROUTE_NAMES = [
+        "None",
+        "Mean",
+        "Spread",
+        "Jitter",
+        "Bin"
+    ]
+
+    def __init__(self):
+        super().__init__()
+
+        self.outputs = [
+            DelayedOutput(cv4, cv1),
+            DelayedOutput(cv5, cv2),
+            DelayedOutput(cv6, cv3)
+        ]
+
+        ## Voltage bins for bin mode
+        self.voltage_bins = [
+            ContinuousBin("Continuous"),
+            VoltageBin("Bin 2", [0, 10]),
+            VoltageBin("Bin 3", [0, 5, 10]),
+            VoltageBin("Bin 6", [0, 2, 4, 6, 8, 10]),
+            VoltageBin("Bin 7", [0, 1.7, 3.4, 5, 6.6, 8.3, 10]),
+            VoltageBin("Bin 9", [0, 1.25, 2.5, 3.75, 5, 6.25, 7.5, 8.75, 10])
+        ]
+
+        # create bins for the quantized 1V/oct modes
+        VOLTS_PER_TONE = 1.0 / 6
+        VOLTS_PER_SEMITONE = 1.0 / 12
+        VOLTS_PER_QUARTERTONE = 1.0 / 24
+        tones = []
+        semitones = []
+        quartertones = []
+        for oct in range(10):
+            for tone in range(6):
+                tones.append(oct + VOLTS_PER_TONE * tone)
+
+            for semitone in range(12):
+                semitones.append(oct + VOLTS_PER_SEMITONE * semitone)
+
+            for quartertone in range(24):
+                quartertones.append(oct + VOLTS_PER_QUARTERTONE * quartertone)
+
+        self.voltage_bins.append(VoltageBin("Tone", tones))
+        self.voltage_bins.append(VoltageBin("Semitone", semitones))
+        self.voltage_bins.append(VoltageBin("Quartertone", quartertones))
+
+        cfg = self.load_state_json()
+
+        self.mean = cfg.get("mean", 0.5)
+        self.stdev = cfg.get("stdev", 0.5)
+        self.ain_route = cfg.get("ain_route", 0)
+        self.voltage_bin = cfg.get("bin", 0)
+        self.jitter = cfg.get("jitter", 0)
+
+        # create the lockable knobs
+        #  Note that this does mean _sometimes_ you'll need to sweep the knob all the way left/right
+        #  to unlock it
+        self.k1_bank = (
+            KnobBank.builder(k1)
+            .with_unlocked_knob("mean")
+            .with_locked_knob("jitter", initial_percentage_value=cfg.get("jitter", 0.5))
+            .build()
+        )
+        self.k2_bank = (
+            KnobBank.builder(k2)
+            .with_unlocked_knob("stdev")
+            .with_locked_knob("bin", initial_percentage_value=int(self.voltage_bin / len(self.voltage_bins)))
+            .build()
+        )
+
+        self.config_dirty = False
+        self.output_dirty = False
+
+        self.last_interaction_at = time.ticks_ms()
+        self.screensaver = Screensaver()
+
+        self.last_clock_at = time.ticks_ms()
+
+        self.clock_duration_ms = 0
+        self.clock_duty_cycle_ms = 0
+
+        @b1.handler
+        def on_b1_rise():
+            self.k1_bank.next()
+            self.k2_bank.next()
+            self.last_interaction_at = time.ticks_ms()
+
+        @b1.handler_falling
+        def on_b1_fall():
```

Comment:
[required] This looks like it will change the knob bank handler on _both_ rise and fall? Shouldn't the button only change knob bank on one button direction? B2 changes AIN destination on rise, so I assume you intend to attach B1 behavior on rise alone too.

Reply:
That's correct; the intent is that the button acts as a momentary shift button. It should be held down to allow changing the binning mode or the jitter.  I've changed the code slightly to make it clearer what's going on, but the intent is that the knob bank handler _should_ be set on both the rising & falling edge in order that `b1` act as a momentary button instead of a latched one.